src/main/kotlin/com/hibob/types/StoreService.kt

```diff
@@ -0,0 +1,65 @@
+package com.hibob.types
+
+import org.springframework.stereotype.Component
+import java.time.LocalDate
+
+@Component
+class StoreService {
+
+    fun checkProductCustom(custom: Any): Boolean{
+        (custom as? Boolean)?.let{
+            when(custom){
+                true -> return true
+                else -> return false
+            }
+        } ?: return false
+    }
+
+    fun fail(message: String): Nothing {
+        throw IllegalStateException(message)
+    }
+
+    fun calculateTotal(products: List<Product>): Double {
+        return products.map {it.price}.sum()
+    }
+
+    fun findStatus(payment: Payment, total: Double): Statuses {
+        return when(payment){
+            is Payment.Cash -> fail("Can't pay with cash")
+
+            is Payment.CreditCard -> {
+                if(payment.number.length == 10 &&
```

Comment:
this condition is not clear first I suggest you in case of CreditCard to call function called validateCreditCard.
then I would suggest you to split the condition to parts, and each part to use function that going to be more readable and easier to fix.

something like:
```
  fun checkPayment (payment: Payment ,price : Double): Statuses {
        return when(payment){
            is Payment.Cash-> fail("you cant pay with cash")
            is Payment.PayPal-> {
                if(payment.email.contains('@'))
                    Statuses.SUCCESS
                else
                    Statuses.FAILURE
            }
            is Payment.CreditCard-> if(isCreditCardValid(payment, price)) Statuses.SUCCESS else Statuses.FAILURE
        }
    }

    private fun isCreditCardValid(payment: Payment.CreditCard, price: Double) =
        isSupportedType(payment.type) && isCardDetailsValid(payment) && isNotOverTheLimit(price, payment)

    private fun isSupportedType(cardType: CreditCardType) =
        (cardType == CreditCardType.MASTERCARD || cardType == CreditCardType.VISA)

    private fun isCardDetailsValid(payment: Payment.CreditCard) =
        payment.expiryDate.isAfter(LocalDate.now())  && payment.number.length == 10

    private fun isNotOverTheLimit(price: Double, payment: Payment.CreditCard) = price < payment.limit

```
look at the condition now, I think its easier to read it

Reply:
Thanks!

Instead of this change: 
is Payment.CreditCard-> if(isCreditCardValid(payment, price)) Statuses.SUCCESS else Statuses.FAILURE

Can I write:
is Payment.CreditCard -> validateCreditCard(payment)

the function validateCreditCard will return type Statuses.

Or is that an incorrect way to write this?

Reply:
yes, you can do it :)